### DISCUSSION TITLE
How can I specify the output directory for native runtime libraries?

### DISCUSSION THREAD

**Discussion by sofakng** [NuGet/Home#12900] [Q&A]
I have a native library with many files:

- thunk64to32\pixdfltn.dll
- thunk64to32\pixmdln.dll
- ...

I have them in this folder structure to build the NuGet package:

- \runtimes\win-x64\lib\native\thunk64to32\*.*

However, they don't get included in the build folder of my application that uses this library.  I think I can use copy them to the \native folder but then they will get output into the base output folder of my application.

I want these to be output in a subfolder of my application.

Example:  MyApplication\bin\release\thunk64to32

Does that make sense?  Thanks for any help!

**Comment by zivkan** [comment]
Native C++ projects don't support PackageReference, and `runtimes/` files in NuGet packages is a PackageReferene only feature. To support vcxproj C++ projects, I believe you'll need to ship your own `build/native/PackageId.targets` file and interact with the C++ build system "manually".

If you're creating a package for .NET projects, then I have this draft docs which explains this topic: https://github.com/NuGet/docs.microsoft.com-nuget/pull/3092

Because I'm "letting perfect be the enemy of good" (I don't want to publish docs with factual errors), it takes a lot of effort to test both SDK and non-SDK style .NET projects, which is the main reason I haven't finished it and merged it yet.

Anyway, the naming convention for the directories is not correct. `runtimes/{rid}/lib/{tfm}/*` is used, but being a .NET project, this will be something like `runtimes/win-x64/lib/net6.0/something.dll`. The directory structure you used would only be applicable for project whose TargetFramework is `native`, which is actually what NuGet uses for C++ projects, but as mentioned this is a PackageReference feature that vcxproj project don't support. For packages aimed at .NET projects, the native libraries (targets of P/Invoke) go in `runtimes/{rid}/native/*` (no `lib/`).
